internal/jobtracker/jobtracker.go

```diff
@@ -0,0 +1,134 @@
+// That is, their status can be checked and they can be stopped.
+package jobtracker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/juju/zaputil/zapctx"
+	"go.uber.org/zap"
+)
+
+// Store defines the interface for tracking the lifecycle and status of jobs.
+// It provides methods to add a new job, update its status (running, successful, or failed),
+// and retrieve a stop signal for a specific job.
+type Store interface {
+	AddJob(ctx context.Context, jobType string) (uuid.UUID, error)
+	SetJobRunning(ctx context.Context, jobId uuid.UUID) error
+	SetJobSuccessful(ctx context.Context, jobId uuid.UUID) error
+	SetJobFailed(ctx context.Context, jobId uuid.UUID, jobErr error) error
+	GetJobStopSignal(ctx context.Context, jobId uuid.UUID) (stopSignal bool, err error)
+}
+
+// Tracker manages job tracking operations using a provided JobTrackerStore.
+// It periodically performs tasks based on the specified stopInterval duration.
+type Tracker struct {
+	store        Store
+	stopInterval time.Duration
+}
+
+// NewJobTracker creates and returns a new Tracker instance using the provided JobTrackerStore and stopInterval.
+// It returns an error if the store is nil or if stopInterval is not greater than zero.
+func NewJobTracker(store Store, stopInterval time.Duration) (*Tracker, error) {
+	if store == nil {
+		return nil, errors.New("store cannot be nil")
+	}
+	if stopInterval <= 0 {
+		return nil, errors.New("stopInterval must be greater than zero")
+	}
+
+	return &Tracker{
+		stopInterval: stopInterval,
+		store:        store,
+	}, nil
+}
+
+// Run runs a new job and returns the job ID.
+func (j *Tracker) Run(ctx context.Context, jobType string, job func(ctx context.Context) error, maxDuration time.Duration) (uuid.UUID, error) {
+	jobId, err := j.store.AddJob(ctx, jobType)
+	if err != nil {
+		return jobId, err
+	}
+
+	go j.handleJob(ctx, jobId, maxDuration, job)
+
+	return jobId, nil
+}
+
+// handleJob runs a job with a given context, job ID, and deadline.
+// It manages the job's lifecycle, including setting its status in the store, handling retries on store operations,
+// and responding to stop signals or context cancellations. The job is run in a separate goroutine, and its status
+// is updated as running, successful, or failed based on its result or context expiration.
+// If a stop signal is received or the context is canceled or times out, the job is marked as failed.
+// Store operations are retried up to 5 times with a 30-second delay between attempts in case of transient errors.
+func (j *Tracker) handleJob(
+	ctx context.Context,
+	id uuid.UUID,
+	maxDuration time.Duration,
+	job func(ctx context.Context) error,
+) {
+	jobCtx, cancelJob := context.WithTimeout(ctx, maxDuration)
+	defer cancelJob()
+
+	jobErrCh := make(chan error)
+
+	go j.runJob(ctx, jobCtx, id, jobErrCh, job)
+	j.monitorJob(ctx, jobCtx, id, jobErrCh, cancelJob)
```

Comment:
Why do we have 2 contexts?

Reply:
One is the parent context, one is the context for the running job. The parent context is simply used for logging, and the job context is used for routine tracking.

Reply:
I don't think I'm following why we would need the parent context for logging.
I think you should delete the parent context and just use the one context that is created in `handleJob`